### PR TITLE
Add Reveal attachment generator

### PR DESCRIPTION
**Pull request by NickEntin**
This introduces a new AardvarkReveal framework that contains utilities for generating a bug report attachment containing a [Reveal](https://revealapp.com/) bundle.

### CODE REVIEW

File: Sources/AardvarkReveal/RevealServiceBrowser.swift
Comment (NickEntin) [inline Sources/AardvarkReveal/RevealServiceBrowser.swift:31]: This tries to find a Reveal server running on the same device. It doesn't differentiate between multiple Reveal servers running on the same device, which may cause unexpected behavior when running apps in multiple simulators. That said, I don't think filing a bug report from a simulator is a common use case currently, so this probably isn't something we need to worry about immediately.

File: Sources/AardvarkReveal/ARKCompressionUtility.h
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKCompressionUtility.h:1]: I'm tempted to put this utility into the `Aardvark` module, since it isn't specific to the Reveal attachment generation. I could see it being useful for other attachments as well (even being able to compress the log stores, to reduce the data size of bug reports).

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: I'd really rather this not be part of the public API, but [from what I can tell that's not easily done](https://medium.com/5-minute-break-while-coding/creating-swift-framework-with-private-objective-c-members-the-good-the-bad-and-the-ugly-4d726386644b). Hopefully the class name and headerdoc comment make it clear that this is _not_ appropriate for use as a general purpose archival tool.

One day it may make sense to build this out further and move it into the `Aardvark` module, since it could be useful for other types of attachments that have a similar bundle structure, but there's a lot of limitations of it right now that I'd rather wait until we have a real use case to deal with.

**Comment by dfed** [comment]
Tag me when this is ready for review (it's still a draft)

**Comment by NickEntin** [comment]
@dfed Still looking into a couple things, but I think this is in an okay place to review for landing as-is and iterate on it in follow up PRs. Will comment on the places I'm thinking about improving.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:262]: I'd feel more confident reviewing this code if this code had unit tests.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: what are these magic numbers? the 7 and 3? Can we make them named constants?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: I wonder if we can get this to communicate with the Reveal server via the `NetService` directly, rather than HTTP requests, to avoid requiring insecure HTTP requests to `localhost`.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (dfed) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:267]: Can these argument names be made to be human-readable (longer than 1 character)?

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:19]: `///` comment?

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:60]: `///` comment?

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [inline Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:53]: I want to follow up here with having this write the archive data to disk, rather than storing it in memory. As is, this could potentially cause some OOM problems during the attachment generation. The complexity of that comes on the compression side, where it needs to page in data from disk while it's compressing it.

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:86]: 😬 can we call this `className`? If not, why not?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:130]: one hell of a comment. well done.

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:161]: 😬 why are we kicking off work from `init`?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:206]: I thought our minimum version was iOS 12?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:220]: can we turn `-1` into a named constant?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:457]: could this method be made static?

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: 50 is significantly lower than 100 😅

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [inline Sources/AardvarkReveal/RevealAttachmentGenerator.swift:481]: I'm struggling with how this class can be modified confidently in the future. Can we add some unit or contractural tests?

**Review by dfed — COMMENTED** [review]
Solid stuff! Lots of thoughts, but we're getting somewhere 🙂

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: It is indeed. The longest file path right now is for the resource images:

```
<app_name>.reveal/Resources/2692677334741286913#1.png
```

That's the app name, followed by ".reveal/Resources/", a numeric identifier, "#", either "0" or "1", and the file extension. I'm assuming the identifiers are always 19 digits, although I'm not 100% sure of that. That gives us 43 characters plus a null terminator. So in theory this limit could be 56 characters, but 50 felt like a reasonable limit that gives a little bit of padding in case I'm wrong about the length of the identifier being capped at 19.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: This is the conversion to octal (base 8). Doing a bitwise and with 7 gets you the last three bits (`7` in decimal is `111` in binary). Add that to `'0'` and you get that number in ASCII (`'0'` through `'7'`). Then you shift by 3 bits to get the next least significant digit.

Maybe it would be clearer to replace `7` with `0b111`?

File: AardvarkSample/AardvarkSample/AppDelegate.swift
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:112]: Correct. Or, at least, it's definitely a bug if it isn't.

File: .github/workflows/ci.yml
Comment (NickEntin) [reply .github/workflows/ci.yml:56]: Great question. I think in the long term it definitely makes sense to grow out an ecosystem of components in separate repos, with this repo holding the core components and maybe a link to the other repos in the README. The main reason for having it here for now is to have it in the sample app. Maybe it's still worth starting that pattern now and moving it to a separate repo with its own sample app...

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (dfed) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:315]: Got it. The file path limit is def odd. How you're tackling this makes sense.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (dfed) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:278]: `0b111` would be a bit more clear. I'd also strongly recommend wring tests for this code.

Comment (sethfri) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: I'm just a fly on the wall, but I think this could also be made clearer by taking the explanation you've provided here on GitHub and turning that into more granular comments (rather than one comment at the start of this code block). That way, you're walking the reader through it a bit more, and readers will have to brush up less on their bit manipulation 🙏🏻

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: Correct, this problem would go away if this were written in Swift. I find it a lot easier to work with things that require specific memory layout (the header struct in particular here) in Objective-C than Swift though. If you have suggestions for how this could be ported to Swift without requiring a bunch of unsafe pointers, I'm definitely open to it.

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:130]: Thanks, this took quite a while to write up 😂

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:161]: Fair question. I was trying to make it hard to get wrong (i.e. if you forget to start it early). Maybe that's too magical and I should expose the start/stop methods.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.h
Comment (dfed) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.h:23]: 😂 I mean... yeah it's certainly harder to write this kind of code in Swift. I don't have much in the way of suggestions here.

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: Addressed in 462da40

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:252]: Added some tests in 462da40

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m:269]: Addressed in 462da40 and aa78c12

File: Sources/AardvarkReveal/ARKRevealArchiveBuilder.m
Comment (NickEntin) [reply Sources/AardvarkReveal/ARKRevealArchiveBuilder.m]: Addressed in 77aec64

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in c528e39

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: @tonyarnold Do you know if communicating via the `NetService` directly would remove the need for this requirement? Or is there an alternate option for communicating with the Reveal server? I don't have much experience with `NetService`, so not sure what options are available.

Comment (tonyarnold) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: My understanding is that `NetService` is only used for broadcasting availability of services and clients of services. You can't communicate over them directly, so as far as I'm aware, Reveal's `NetService` would be broadcasting details of the HTTP endpoint and the port it is running on to net service browsers. 

I'll have a look when I'm back at my desk later today, and verify exactly what we're doing.

Comment (tonyarnold) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: Side note: by _golly_ I wish that we offered alternatives here!

We've been a bit hamstrung in terms of some of the nice new technology Apple has added in recent releases - our support for older iOS releases means that we can't use many of them in the server component. I'll need to reassess that now that we're moving our minimum iOS and tvOS versions up to 10.x releases.

Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:152]: Ahh, makes sense! Sending the requests over HTTP seems to work great with the exception of this requirement. It would be awesome to get a little bit cleaner alternative, but I definitely don't think this would be a blocker.

**Comment by NickEntin** [comment]
Came across the file coordinator's utility for creating a ZIP archive from a directory and switched over to using that in favor of the custom TAR builder. This should resolve a lot of the open comments right now.

File: AardvarkSample/AardvarkSample/AppDelegate.swift
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:109]: - [ ] Add comment about ignoring interaction events while bundle is being generated

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:206]: Not yet, #92 is still open.

- [ ] Remove unnecessary availability checks once deployment target bump lands

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:220]: - [ ] Turn missing port check into named constant

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:475]: - [ ] Add tests for attachment generator

File: AardvarkSample/AardvarkSample/AppDelegate.swift
Comment (NickEntin) [reply AardvarkSample/AardvarkSample/AppDelegate.swift:111]: Addressed in dc6ea26

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in 33d1f73

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift]: Addressed in f9a5614

File: Sources/AardvarkReveal/RevealAttachmentGenerator.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:490]: Added some integration tests for the generator's callbacks in different scenarios. Still want to get one more big one in to test its interaction with the archive builder.

Comment (NickEntin) [reply Sources/AardvarkReveal/RevealAttachmentGenerator.swift:495]: Pretty happy with the test setup I've got now. Probably a few more edge cases that could be covered, but I think it's got pretty good coverage already.

**Comment by NickEntin** [comment]
@dfed I think this is ready for another review pass when you get a chance. I'm looking into moving this extension into its own repo, but I'd like to still go through the review process here and get everything approved before copying it over.

**Comment by NickEntin** [comment]
@dfed New repo is good to go! Once this is PR approved I'll copy the code over to the new repo and make it public.

File: Sources/AardvarkReveal/ARKCompressionUtility.m
Comment (dfed) [reply Sources/AardvarkReveal/ARKCompressionUtility.m:24]: alternatively, why not create a dependency on the GZIP library?

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:63]: If this case name is required due to compatibility with another library or model, would be good to state as much in a comment

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:19]: repeating myself, but `///` comments would be great even on internal classes

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (dfed) [inline Sources/AardvarkReveal/ApplicationState.swift:179]: same thing here. if these shortened coding keys are necessary, let's explain why in a comment

File: Sources/AardvarkRevealTests/ApplicationStateTests.swift
Comment (dfed) [inline Sources/AardvarkRevealTests/ApplicationStateTests.swift:23]: love these test, but naming the test methods in a more elucidating way would be great. Consider writing test methods using the form `test_thingBeingTested_stateExpectation` to make long-term maintenance easier.

**Review by dfed — APPROVED** [review]
LGTM, but `///` comments on internal types would be helpful for future maintainers.

File: Sources/AardvarkReveal/ApplicationState.swift
Comment (NickEntin) [reply Sources/AardvarkReveal/ApplicationState.swift:70]: Added a comment at the top of the file. This whole file mirrors the format used in the JSON representation of the application state included in a `.reveal` file.

**Comment by NickEntin** [comment]
I'm going to close this PR now. This source code will all be copied over to a new repo as-is, along with a new demo app and documentation. I'll post back here with a link to the new repo once it's live. Thanks for the reviews!

**Comment by dfed** [comment]
Just fat-fingered that deletion! Giant touch target on iPad 🤦‍♂️

**Pull request closed without merging**
